File: shared/types/summary.ts
import type { FailureReason, RunStatus } from "./contracts";

export type SamplingAggregateMetric = {
  mean: number | null;
  stddev: number | null;
  variance: number | null;
  min: number | null;
  max: number | null;
};

export type RunSamplingSummary = {
  group_id: string;
  sample_index: number;
  sample_count: number;
  completed_sample_count: number;
  sample_run_ids: string[];
  consistency_rate: number | null;
  success_rate: SamplingAggregateMetric;
  average_latency_ms: SamplingAggregateMetric;
  total_cost: SamplingAggregateMetric;
};

export type DashboardCategorySummary = {
  category: string;
  total_tasks: number;
  successful_tasks: number;
  failed_tasks: number;
  success_rate: number;
  average_latency_ms: number | null;
  total_cost: number;
};

export type DashboardFailureSummary = {
  failure_reason: FailureReason;
  count: number;
};

export type DashboardFailedCase = {
  task_run_id: string;
  run_id: string;
  dataset_item_id: string;
  category: string;
  failure_reason: FailureReason;
  error_message: string | null;
  final_output: string | null;
};

export type RunDashboardSummary = {
  run_id: string;
  agent_version_id: string;
  dataset_id: string;
  scorer_config_id: string;
  status: RunStatus;
  total_tasks: number;
  successful_tasks: number;
  failed_tasks: number;
  review_needed_count: number;
  success_rate: number;
  average_latency_ms: number | null;
  total_cost: number;
  sampling: RunSamplingSummary | null;
  category_breakdown: DashboardCategorySummary[];
  failure_breakdown: DashboardFailureSummary[];
  failed_cases: DashboardFailedCase[];
};

File: backend/app/schemas/summary.py
from __future__ import annotations

from pydantic import BaseModel

from app.schemas.contracts import FailureReason, RunStatus


class SamplingAggregateMetricSchema(BaseModel):
    mean: float | None = None
    stddev: float | None = None
    variance: float | None = None
    min: float | None = None
    max: float | None = None


class RunSamplingSummarySchema(BaseModel):
    group_id: str
    sample_index: int
    sample_count: int
    completed_sample_count: int
    sample_run_ids: list[str]
    consistency_rate: float | None = None
    success_rate: SamplingAggregateMetricSchema
    average_latency_ms: SamplingAggregateMetricSchema
    total_cost: SamplingAggregateMetricSchema


class DashboardCategorySummarySchema(BaseModel):
    category: str
    total_tasks: int
    successful_tasks: int
    failed_tasks: int
    success_rate: float
    average_latency_ms: float | None = None
    total_cost: float = 0.0


class DashboardFailureSummarySchema(BaseModel):
    failure_reason: FailureReason
    count: int


class DashboardFailedCaseSchema(BaseModel):
    task_run_id: str
    run_id: str
    dataset_item_id: str
    category: str
    failure_reason: FailureReason
    error_message: str | None = None
    final_output: str | None = None


class RunDashboardSummarySchema(BaseModel):
    run_id: str
    agent_version_id: str
    dataset_id: str
    scorer_config_id: str
    status: RunStatus
    total_tasks: int
    successful_tasks: int
    failed_tasks: int
    review_needed_count: int
    success_rate: float
    average_latency_ms: float | None = None
    total_cost: float = 0.0
    sampling: RunSamplingSummarySchema | None = None
    category_breakdown: list[DashboardCategorySummarySchema]
    failure_breakdown: list[DashboardFailureSummarySchema]
    failed_cases: list[DashboardFailedCaseSchema]
